### PR TITLE
Generalize case where only 1 particle of each type is present

### LINKED ISSUE #3137: Two body Jastrow issue when there are more than 2 particle types but only one particle of each type
**Describe the bug**
Thanks to @markdewing, I was able to progress further until I broke things again 😄 

When running a calculation with 1 particle in each of >2 particle types, there is a segfault.

**To Reproduce**
1. b1a4692e86ef7b2a55242f3ba845d7eb88191ad9-dirty
2. cmake command : `cmake -DCMAKE_BUILD_TYPE=Debug ..`
3. `qmcpack H_wpos.qmc.in-wfj.xml`
The files needed to replicate this issue are located here https://gist.github.com/shivupa/e390aa40b77d6957d72f59b3c1be7203.

**Expected behavior**
The calculation wouldn't segfault. I think the two body jastrow builder should bypass building the uu jastrow?

**System:**
 - Workstation Intel i7-2600S

**Additional context**
Potential Solution?:
[src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h 94](https://github.com/QMCPACK/qmcpack/blob/2271e10df29a71705087820065e5b32b57d90b75/src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h#L94) seems to guard against the 1 elec up/1 elec down case. This might be the source of the issue.

Portion of Output before Seg Fault:
```
========================================================
--- Memory usage report : Memory Usage after cloning ---
========================================================
Available memory on node 0, free + buffers :    6696 MiB
Memory footprint by rank 0 on node 0       :      25 MiB
========================================================
  Initial partition of walkers 0 1 2 3 4 5 6 7 8 

  Using Particle by Particle moves
  Walker moves without drift
  Total Sample Size   =0
  Walker distribution on root = 0 1 2 3 4 5 6 7 8 
[1]    8353 segmentation fault (core dumped)  ~/Documents/qmcpack/build/bin/qmcpack H_wpos.qmc.in-wfj.xml
```

gdb trouble shooting
```

Thread 1 "qmcpack" received signal SIGSEGV, Segmentation fault.
0x00005555558ee984 in qmcplusplus::BsplineFunctor<double>::evaluateVGL (this=0x0, iat=0, iStart=0, 
    iEnd=0, _distArray=0x555556157e20, _valArray=0x55555612bc40, _gradArray=0x55555612bcc0, 
    _laplArray=0x55555612bd40, distArrayCompressed=0x55555602dfe0, distIndices=0x555556128720)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/BsplineFunctor.h:687
687	  real_type dSquareDeltaRinv = DeltaRInv * DeltaRInv;
(gdb) bt
#0  0x00005555558ee984 in qmcplusplus::BsplineFunctor<double>::evaluateVGL (this=0x0, iat=0, 
    iStart=0, iEnd=0, _distArray=0x555556157e20, _valArray=0x55555612bc40, _gradArray=0x55555612bcc0, 
    _laplArray=0x55555612bd40, distArrayCompressed=0x55555602dfe0, distIndices=0x555556128720)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/BsplineFunctor.h:687
#1  0x00005555558e5a3d in qmcplusplus::J2OrbitalSoA<qmcplusplus::BsplineFunctor<double> >::computeU3 (
    this=0x55555612b660, P=..., iat=0, dist=warning: RTTI symbol not found for class 'qmcplusplus::Vector<double, qmcplusplus::Mallocator<double, 32ul> >'
..., u=0x55555612bc40, du=0x55555612bcc0, 
    d2u=0x55555612bd40, triangle=true)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/J2OrbitalSoA.h:485
#2  0x00005555558cd900 in qmcplusplus::J2OrbitalSoA<qmcplusplus::BsplineFunctor<double> >::recompute (
    this=0x55555612b660, P=...)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/J2OrbitalSoA.h:605
#3  0x00005555558cee34 in qmcplusplus::J2OrbitalSoA<qmcplusplus::BsplineFunctor<double> >::evaluateGL
    (this=0x55555612b660, P=..., G=warning: RTTI symbol not found for class 'qmcplusplus::ParticleAttrib<qmcplusplus::TinyVector<double, 3u>, std::allocator<qmcplusplus::TinyVector<double, 3u> > >'
..., L=..., fromscratch=true)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/J2OrbitalSoA.h:662
#4  0x00005555558cd798 in qmcplusplus::J2OrbitalSoA<qmcplusplus::BsplineFunctor<double> >::evaluateLog
    (this=0x55555612b660, P=..., G=warning: RTTI symbol not found for class 'qmcplusplus::ParticleAttrib<qmcplusplus::TinyVector<double, 3u>, std::allocator<qmcplusplus::TinyVector<double, 3u> > >'
..., L=...)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/Jastrow/J2OrbitalSoA.h:652
#5  0x0000555555861c01 in qmcplusplus::TrialWaveFunction::evaluateLog (this=0x5555560274e0, P=...)
    at /home/shiv/Documents/qmcpack/src/QMCWaveFunctions/TrialWaveFunction.cpp:119
```

I didn't find this incredibly enlightening besides this resulting in a divide by zero.

Let me know what you think.

### PR DESCRIPTION
**Pull request by shivupa**
## Proposed changes
Closes #3137 
This implements the fixes @markdewing recommended. It generalizes the special case of the two body jastrow where there are only 1 particle of each type present.
## What type(s) of changes does this code introduce?
- Bugfix

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Ryzen workstation (5900x, gcc version 10.2.0)
## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
OK to test

**Review by ye-luo — CHANGES_REQUESTED** [review]
Please add a unit test, The J2 in `opt/qmcpack/src/QMCWaveFunctions/tests/test_bspline_jastrow.cpp` can be a starting point.

**Review by prckent — COMMENTED** [review]
Many thanks for this Shiv. FYI I ran the full (but limited) set of optimzation tests and they seemed to be OK.

I agree that adding a unit test by expanding the example Ye pointed to would be good.

If you don't have time to do that let us know.

**Comment by shivupa** [comment]
Yup sounds good. Thanks for the example Ye. I'll give it a try and report back

**Comment by markdewing** [comment]
The tests in `test_DiffTwoBodyJastrowOrbital.cpp` would be a good starting point, since they target the previous round of problems in this code.

**Comment by ye-luo** [comment]
The added test doesn't trigger failure on develop.
The test is named as "DiffTwoBodyJastrowOrbital" but there is no call to routines calculating parameter derivatives.
Please update the test.

**Comment by shivupa** [comment]
Thanks working on this now. I think I misunderstood what was going on and Mark's previous test. Sorry about that

**Comment by shivupa** [comment]
Sorry is it possible to stop checks on this? I though converting to draft would, but I this these will pass regardless as the previous work did. I mainly pushed to ask some questions.

**Comment by prckent** [comment]
We can't straightforwardly stop tests on a per-PR basis. However, we ignore the test results until the PR is marked as ready.

**Comment by shivupa** [comment]
Makes sense. I didn't want to waste the computational time. 

This is ready for review now! Thanks @ye-luo and @markdewing for the discussion. The test now checks the F vector for any nullptrs. This fails on develop as expected (the F vector is private on develop so once the function allowing access for F is added on develop, this will still fail as expected).

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h [hunk 1]
```diff
@@ -50,6 +50,8 @@ class DiffTwoBodyJastrowOrbital : public DiffWaveFunctionComponent
   std::map<std::string, FT*> J2Unique;
 
 public:
+  // return for testing
+  std::vector<FT*>& getFvector() { return F; }
```
Comment (markdewing) [inline src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h:54]: A more descriptive name would be nice - maybe getPairFunctions?

File: src/QMCWaveFunctions/tests/test_DiffTwoBodyJastrowOrbital.cpp [hunk 2]
```diff
@@ -287,4 +287,77 @@ TEST_CASE("DiffTwoBodyJastrowOrbital other variables", "[wavefunction]")
...
+  // currently opposite spions won't be set to be equivalent
+  // setting u,p doesn't set d,p
+  jorb.addFunc(1, 2, &j2b);
+
+  opt_variables_type global_active;
+  global_active.insertFrom(j2a.myVars);
+  global_active.insertFrom(j2b.myVars);
+  global_active.resetIndex();
+
+  jorb.checkOutVariables(global_active);
```
Comment (markdewing) [inline src/QMCWaveFunctions/tests/test_DiffTwoBodyJastrowOrbital.cpp:356]: I think this code can be removed.

File: src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h (hunk 1 above)
Comment (shivupa) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h]: Fixed!

Comment (shivupa) [reply src/QMCWaveFunctions/Jastrow/DiffTwoBodyJastrowOrbital.h]: Fixed

**Review by markdewing — DISMISSED** [review]
Looks good. Thanks for doing this!

**Review by ye-luo — APPROVED** [review]
Thank you!!!!

**Pull request merged**